File: python/revo3/touch_hybrid.py
"""Verification script for Revo3 Fingertip Force/Torque + Pressure Array Hybrid Touch sensors.

Verifies:
1. Touch layout configuration: 5 Fingertip Force/Torque fingertips (3D force/torque + 48 pts) + 5 Pressure Array fingerpads (57/52 pts) + 1 Pressure Array palm (36 pts).
2. Live hybrid touch snapshot reading.
3. Fingertip Force/Torque 5-fingertip force & torque vector parsing (Fx, Fy, Fz, Tx, Ty, Fn).
4. Pressure Array 6-module tactile array reading with runtime ADC/mN unit detection.
5. Module enable state querying.
6. Zero / Tare calibration execution and verification.
7. Legacy read mode switching (PointArray <-> LegacyForceSummary).
"""

import argparse
import asyncio
import sys

from bc_revo3_sdk import main_mod as sdk


def enum_name(value) -> str:
    """Return the variant name for a PyO3 enum or a regular Python enum."""
    name = getattr(value, "name", None)
    if name:
        return name
    return str(value).rsplit(".", 1)[-1]


def protocol_type(name: str):
    """Resolve the protocol enum across SDK 2.x binding names."""
    enum_type = sdk.ProtocolType
    return getattr(enum_type, name)


def touch_read_mode(*names: str):
    """Resolve TouchReadMode names across SDK 2.0 RC bindings."""
    for name in names:
        member = getattr(sdk.TouchReadMode, name, None)
        if member is not None:
            return member
    raise RuntimeError(f"TouchReadMode does not provide any of: {', '.join(names)}")


def build_force_torque_pressure_array_layout() -> sdk.TouchLayout:
    """Build canonical Fingertip Force/Torque + Pressure Array hybrid touch layout."""
    fingertip_force_torque_signals = [
        sdk.TouchSignal.TouchPoint,
        sdk.TouchSignal.Force3D,
        sdk.TouchSignal.Torque2D,
        sdk.TouchSignal.ResultantForce,
    ]
    modules = []
    # 5 Fingertip Force/Torque Fingertips (0..4)
    for i in range(5):
        modules.append(
            sdk.TouchModuleLayout(
                "fingertip_force_torque_48",
                i * 2 + 1,
                sdk.TouchRegion.Fingertip,
                i,
                fingertip_force_torque_signals,
                48,
            )
        )
    # 5 Pressure Array FingerPads (5..9): Thumb (57), Index..Pinky (52)
    pressure_array_pad_counts = [57, 52, 52, 52, 52]
    for i, count in enumerate(pressure_array_pad_counts):
        layout_id = "pressure_array_thumb_pad_57" if i == 0 else "pressure_array_finger_pad_52"
        modules.append(
            sdk.TouchModuleLayout(
                layout_id,
                    (i + 1) * 2,
                sdk.TouchRegion.FingerPad,
                i,
                [sdk.TouchSignal.TouchPoint],
                count,
            )
        )
    # 1 Pressure Array Palm (10): 36 points
    modules.append(
        sdk.TouchModuleLayout(
            "pressure_array_palm_36",
            0,
            sdk.TouchRegion.Palm,
            0,
            [sdk.TouchSignal.TouchPoint],
            36,
        )
    )
    return sdk.TouchLayout(modules)


def parse_args():
    parser = argparse.ArgumentParser(description="Verify Revo3 Fingertip Force/Torque + Pressure Array Hybrid Touch")
    parser.add_argument("--protocol", choices=["canfd", "modbus", "auto"], default="canfd")
    parser.add_argument("--port", default=None, help="Port name (e.g. brainco:0 or /dev/ttyUSB0)")
    parser.add_argument("--slave-id", type=lambda x: int(x, 0), default=None, help="Slave ID (e.g. 0x7F or 127)")
    parser.add_argument("--count", type=int, default=5, help="Number of telemetry frames to read")
    parser.add_argument("--interval", type=float, default=0.1, help="Polling interval in seconds")
    parser.add_argument("--test-tare", action="store_true", help="Execute touch tare verification")
    parser.add_argument(
        "--test-modes",
        action="store_true",
        help="Test PointArray and the legacy secondary-calibrated summary mode",
    )
    return parser.parse_args()


async def run(args):
    protocol = None
    if args.protocol == "canfd":
        protocol = protocol_type("CanFd")
    elif args.protocol == "modbus":
        protocol = protocol_type("Modbus")

    manager = sdk.Manager()
    try:
        print("=" * 70)
        print("Revo3 Fingertip Force/Torque + Pressure Array Hybrid Touch Verification")
        print("=" * 70)
        print(f"Connecting to Revo3 device (protocol={args.protocol}, port={args.port}, slave_id={args.slave_id})...")

        hand = await manager.connect_auto(
            port=args.port,
            protocol=protocol,
            slave_id=args.slave_id,
            broadcast=False,
        )
        info = hand.device_info
        print(f"\n[OK] Connected: {info.serial_number} ({enum_name(info.hand_side)}) | Model: {enum_name(info.model)}")
        endpoint = getattr(hand, "endpoint", None)
        if endpoint is not None:
            print(f"     Transport: {enum_name(endpoint.protocol)} on {endpoint.port_name} (Slave ID: 0x{endpoint.slave_id:02X})")

        # 1. Apply Fingertip Force/Torque + Pressure Array Layout
        print("\n--- 1. Applying Fingertip Force/Torque + Pressure Array Hybrid Layout ---")
        force_torque_pressure_array_layout = build_force_torque_pressure_array_layout()
        await hand.touch.set_layout(force_torque_pressure_array_layout)
        active_layout = hand.touch.layout
        print(f"[OK] Layout registered: modules = {len(active_layout.modules)}")
        for reg in active_layout.regions:
            print(f"     Region {enum_name(reg.region)}: {len(reg.module_ids)} modules, IDs = {reg.module_ids}")

        # 2. Query Module Enable States
        print("\n--- 2. Checking Module Enable States ---")
        mask = await hand.touch.enabled_mask()
        print(f"[OK] Module enabled bitmask: 0x{mask:03X} (binary: {bin(mask)})")
        for module in active_layout.modules:
            module_id = module.module_id
            en = (mask & (1 << module_id)) != 0
            name = module.layout_id.replace("_", " ")
            print(f"     Module {module_id:2d} ({name:18s}): {'ENABLED' if en else 'DISABLED'}")

        # 3. Read frames using the device's current read mode.
        current_read_mode = await hand.touch.read_mode()
        current_value_mode = await hand.touch.value_mode()
        point_unit = "mN" if int(current_value_mode) == 2 else "ADC"
        print(
            f"\n--- 3. Reading {args.count} Live Hybrid Touch Snapshots "
            f"({enum_name(current_read_mode)}, {enum_name(current_value_mode)}) ---"
        )

        finger_names = ["Thumb", "Index", "Middle", "Ring", "Pinky"]

        for frame_idx in range(1, args.count + 1):
            frame = await hand.touch.snapshot()
            timestamp = getattr(frame, "timestamp", None)
            timestamp_s = (timestamp.sec + timestamp.nsec / 1e9) if timestamp else 0.0
            print(f"\n[Frame #{frame_idx:02d}] Timestamp: {timestamp_s:.3f}s | Sequence: {frame.sequence}")
            raw_modules = list(getattr(frame, "modules", []) or [])
            print("  Module slots: " + ", ".join(
                f"id={getattr(m, 'module_id', '?')}:{getattr(m, 'layout_id', '?')}[{len(getattr(m, 'points', []) or [])}]"
                for m in raw_modules
            ))

            # Parse fingertip_force_torque_* fingertips directly from the public TouchFrame model.
            print("  Fingertip Force/Torque Fingertip Modules (3D Force & Torque):")
            fingertip_force_torque_modules = [
                module
                for module in raw_modules
                if str(getattr(module, "layout_id", "")).startswith("fingertip_force_torque_")
            ]
            for mod in fingertip_force_torque_modules:
                region_index = int(getattr(mod, "region_index", 0))
                fname = finger_names[region_index] if region_index < len(finger_names) else f"Tip {region_index}"
                force = getattr(mod, "force3d", None)
                torque = getattr(mod, "torque2d", None)
                diagnostics = getattr(mod, "diagnostics", None)
                points = list(getattr(mod, "points", None) or [])
                print(
                    f"     [{fname:6s} Tip] module_id={mod.module_id} "
                    f"state={enum_name(mod.sample_state)} "
                    f"sensor_fault={getattr(diagnostics, 'sensor_fault_code_raw', None)!r} | "
                    f"Fx={getattr(force, 'x', 0.0):+8.1f} mN "
                    f"Fy={getattr(force, 'y', 0.0):+8.1f} mN "
                    f"Fz={getattr(force, 'z', 0.0):+8.1f} mN | "
                    f"Mx={getattr(torque, 'x', 0.0):+8.4f} Nm "
                    f"My={getattr(torque, 'y', 0.0):+8.4f} Nm | "
                    f"Fn={getattr(mod, 'resultant_force_mn', 0.0) or 0.0:8.1f} mN | "
                    f"Points={len(points)} (max={max(points) if points else 0})"
                )

            # Parse Pressure Array modules (pads + palm) from the canonical module list.
            print("  pressure_array_* Piezoresistive Array Modules:")
            pressure_array_modules = [
                module
                for module in raw_modules
                if enum_name(getattr(module, "region", "")) in ("Palm", "FingerPad")
            ]
            for module in pressure_array_modules:
                points = list(getattr(module, "points", []) or [])
                if points:
                    active = sum(1 for value in points if value > 0)
                    peak = max(points)
                    average = sum(points) / len(points)
                    print(
                        f"     [id={module.module_id:2d} {module.layout_id:18s}] "
                        f"Points: {len(points):3d} | Peak: {peak:4d} {point_unit} | "
                        f"Mean: {average:6.1f} {point_unit} | Active: {active:3d}/{len(points):3d}"
                    )
                else:
                    regional = getattr(module, "regional_forces_mn", None)
                    print(
                        f"     [id={module.module_id:2d} {module.layout_id:18s}] "
                        f"regional_forces_mn={regional!r} state={enum_name(module.sample_state)}"
                    )

            if frame_idx < args.count:
                await asyncio.sleep(args.interval)

        # 4. Test Zero / Tare if requested
        if args.test_tare:
            print("\n--- 4. Testing Zero / Tare Calibration ---")
            print("Executing global touch tare on all 11 modules...")
            await hand.touch.tare()
            print("[OK] Global tare command sent successfully.")
            await asyncio.sleep(0.1)
            frame_after = await hand.touch.snapshot()
            print("Post-Tare Fingertip Force/Torque Resultant Fn:")
            for mod in frame_after.modules:
                if not str(mod.layout_id).startswith("fingertip_force_torque_"):
                    continue
                region_index = int(mod.region_index)
                fname = finger_names[region_index] if region_index < len(finger_names) else f"Tip {region_index}"
                force = mod.force3d
                print(
                    f"     [{fname:6s}] Fn={mod.resultant_force_mn or 0.0:.1f} mN "
                    f"(Fx={getattr(force, 'x', 0.0):+.1f} mN, "
                    f"Fy={getattr(force, 'y', 0.0):+.1f} mN, "
                    f"Fz={getattr(force, 'z', 0.0):+.1f} mN)"
                )

        # 5. Test Read Modes if requested
        if args.test_modes:
            print(
                "\n--- 5. Testing Legacy Read Mode "
                "(PointArray <-> LegacyForceSummary) ---"
            )
            print("Switching to legacy secondary-calibrated force summary mode...")
            await hand.touch.set_read_mode(
                touch_read_mode("LegacyForceSummary", "ForceSummary")
            )
            await asyncio.sleep(0.05)
            frame_summary = await hand.touch.snapshot()
            regional_values = [
                value
                for module in frame_summary.modules
                for value in (getattr(module, "regional_forces_mn", None) or [])
            ]
            if regional_values:
                print(
                    f"[OK] Legacy summary received {len(regional_values)} regional values: "
                    f"{regional_values[:8]}..."
                )
            else:
                print("[OK] Legacy secondary-calibrated summary mode executed.")

            print("Restoring PointArray mode...")
            await hand.touch.set_read_mode(
                touch_read_mode("PointArray", "TactileArray")
            )
            await asyncio.sleep(0.05)
            print("[OK] PointArray mode restored.")

        print("\n" + "=" * 70)
        print("[SUCCESS] Revo3 Fingertip Force/Torque + Pressure Array Hybrid Touch verification complete!")
        print("=" * 70)

    finally:
        await manager.close()


if __name__ == "__main__":
    asyncio.run(run(parse_args()))

File: c/demo/touch_hybrid.cpp
#include <revo3/revo3.hpp>

#include <chrono>
#include <cstdio>
#include <cstring>
#include <string>
#include <thread>
#include <vector>

static revo3::TouchLayout build_hp_mt_layout() {
  revo3::TouchLayout layout;
  std::vector<revo3::TouchSignal> hp_signals = {
      revo3::TouchSignal::TouchPoint,     revo3::TouchSignal::Force3D,
      revo3::TouchSignal::Torque2D,       revo3::TouchSignal::ResultantForce,
  };
  // 5 HP Fingertips (module_id: 1, 3, 5, 7, 9)
  for (uint8_t i = 0; i < 5; ++i) {
    revo3::TouchModuleLayout mod{};
    mod.layout_id = "hp_fingertip_48";
    mod.module_id = i * 2 + 1;
    mod.region = revo3::TouchRegion::Fingertip;
    mod.region_index = i;
    mod.signals = hp_signals;
    mod.point_count = 48;
    layout.modules.push_back(mod);
  }
  // 5 MT FingerPads (module_id: 2, 4, 6, 8, 10)
  uint16_t mt_pad_counts[5] = {57, 52, 52, 52, 52};
  for (uint8_t i = 0; i < 5; ++i) {
    revo3::TouchModuleLayout mod{};
    mod.layout_id = (i == 0) ? "mt_thumbpad_57" : "mt_fingerpad_52";
    mod.module_id = (i + 1) * 2;
    mod.region = revo3::TouchRegion::FingerPad;
    mod.region_index = i;
    mod.signals = {revo3::TouchSignal::TouchPoint};
    mod.point_count = mt_pad_counts[i];
    layout.modules.push_back(mod);
  }
  // 1 MT Palm (module_id: 0)
  {
    revo3::TouchModuleLayout mod{};
    mod.layout_id = "mt_palm_36";
    mod.module_id = 0;
    mod.region = revo3::TouchRegion::Palm;
    mod.region_index = 0;
    mod.signals = {revo3::TouchSignal::TouchPoint};
    mod.point_count = 36;
    layout.modules.push_back(mod);
  }
  return layout;
}

int main(int argc, char **argv) {
  using namespace std::chrono_literals;
  revo3::init_logging(LOG_LEVEL_INFO, true);

  try {
    revo3::DiscoveryOptions discovery;
    bool test_tare = false;
    for (int index = 1; index < argc; ++index) {
      if (std::strcmp(argv[index], "--test-tare") == 0) {
        test_tare = true;
      } else {
        discovery.port = argv[index];
      }
    }

    std::printf("=================================================================\n");
    std::printf("      Revo3 C++ SDK HP + MT Hybrid Touch Verification Demo       \n");
    std::printf("=================================================================\n");

    revo3::Manager manager;
    auto hand = manager.connect_auto(discovery);
    const auto info = hand.device_info();
    std::printf("[OK] Connected to %s (Model: %u)\n", info.serial_number.c_str(),
                static_cast<unsigned>(info.model));

    auto touch = hand.touch();
    const auto value_mode = touch.value_mode();
    const char *point_unit =
        value_mode == revo3::TouchValueMode::Force ? "mN" : "ADC";

    // 1. Set HP + MT Hybrid Layout
    std::printf("\n--- 1. Setting HP + MT Hybrid Layout (11 Modules) ---\n");
    touch.set_layout(build_hp_mt_layout());
    const auto active_layout = touch.layout();
    std::printf("[OK] Registered layout with %zu modules, %zu regions\n",
                active_layout.modules.size(), active_layout.regions.size());

    // 2. Query Enabled Mask
    std::printf("\n--- 2. Querying Touch Module Enabled Mask ---\n");
    const auto mask = touch.enabled_mask();
    std::printf("[OK] Enabled Mask: 0x%04X\n", mask);

    // 3. Read 3 Snapshots
    std::printf("\n--- 3. Reading 3 Hybrid Touch Snapshots ---\n");
    const char *finger_names[5] = {"Thumb", "Index", "Middle", "Ring", "Pinky"};
    for (int frame_idx = 1; frame_idx <= 3; ++frame_idx) {
      const auto frame = touch.snapshot();
      std::printf("[Snapshot #%02d] Sequence: %llu, Modules: %zu\n", frame_idx,
                  static_cast<unsigned long long>(frame.sequence),
                  frame.modules.size());
      for (const auto &m : frame.modules) {
        if (m.layout_id.rfind("hp_", 0) == 0) {
          const char *fname = (m.region_index < 5) ? finger_names[m.region_index] : "Tip";
          std::printf("  [%-6s Tip] ID=%2u | Fx=%+7.1f Fy=%+7.1f Fz=%+7.1f mN | Mx=%+6.4f My=%+6.4f Nm | Fn=%7.1f mN | Pts=%zu\n",
                      fname, m.module_id,
                      m.has_force3d ? m.force3d.x : 0.0F,
                      m.has_force3d ? m.force3d.y : 0.0F,
                      m.has_force3d ? m.force3d.z : 0.0F,
                      m.has_torque2d ? m.torque2d.x : 0.0F,
                      m.has_torque2d ? m.torque2d.y : 0.0F,
                      m.has_resultant_force ? m.resultant_force_mn : 0.0F,
                      m.points.size());
        } else {
          uint16_t peak = 0;
          for (auto p : m.points) {
            if (p > peak) peak = p;
          }
          std::printf("  [Mod %2u %-18s] Pts=%2zu | Peak=%4u %s\n",
                      m.module_id, m.layout_id.c_str(), m.points.size(), peak,
                      point_unit);
        }
      }
      std::this_thread::sleep_for(50ms);
    }

    // 4. Touch Pull Subscription (20 frames @ 20ms)
    std::printf("\n--- 4. Testing Touch Pull Subscription (20 frames @ 20ms) ---\n");
    auto subscription = touch.subscribe(20ms);
    for (int i = 1; i <= 20; ++i) {
      const auto frame = subscription.next();
      if (i == 1 || i == 10 || i == 20) {
        std::printf("  [Sub Frame #%02d] Seq=%llu, Total Modules=%zu\n", i,
                    static_cast<unsigned long long>(frame.sequence),
                    frame.modules.size());
      }
    }
    subscription.close();
    std::printf("[OK] Subscription closed successfully.\n");

    // 5. Zero / Tare (explicit opt-in because this changes calibration state)
    if (test_tare) {
      std::printf("\n--- 5. Testing Zero / Tare ---\n");
      touch.tare();
      std::printf("[OK] Global tare sent successfully.\n");
      std::this_thread::sleep_for(100ms);
    } else {
      std::printf("\n--- 5. Zero / Tare skipped (pass --test-tare to enable) ---\n");
    }

    const auto stats = hand.statistics();
    std::printf("\n--- 6. Runtime Statistics ---\n");
    std::printf("  Touch reads:       %llu\n", static_cast<unsigned long long>(stats.touch_reads));
    std::printf("  Failed operations: %llu\n", static_cast<unsigned long long>(stats.failed_operations));

    std::printf("\n=================================================================\n");
    std::printf("       [SUCCESS] Revo3 C++ Hybrid Touch Verification Completed!  \n");
    std::printf("=================================================================\n");
    return 0;
  } catch (const revo3::SdkError &error) {
    std::fprintf(stderr, "Revo3 error: %s\n", error.what());
    return 1;
  } catch (const std::exception &error) {
    std::fprintf(stderr, "Error: %s\n", error.what());
    return 1;
  }
}
